[receiver/otlp, internal, featuregate] Add featuregate to switch to localhost defaults for server-like components

mx-psi opened the pull request:
**Description:** 

- Define `component.UseLocalHostAsDefaultHost` in the `internal/localhostgate` package. 
- Define `featuregate.ErrIsAlreadyRegistered` error, returned by `Register` when a gate is already registered.
- Adds support for the localhost gate on the OTLP receiver.

This PR does not remove the current warning in any way, we can remove this separately.

**Link to tracking Issue:** Updates #8510

**Testing:** Adds unit tests

**Documentation:** Document on OTLP receiver template and add related logging.

mx-psi commented:
To see the usage of this API on contrib take a look at open-telemetry/opentelemetry-collector-contrib/pull/27439

iblancasa commented on component/featuregate.go:
```diff
@@ -0,0 +1,40 @@
+// EndpointForPort gets the endpoint for a given port using localhost or 0.0.0.0 depending on the feature gate.
+func EndpointForPort(port int) string {
+	host := "0.0.0.0"
+	if UseLocalHostAsDefaultHostfeatureGate.IsEnabled() {
```
Instead of setting this to `localhost`, would it make sense to allow setting whatever value from the feature flag?

mx-psi replied:
The feature gate is meant to ease migration to the new defaults, but I wouldn't want people relying on it to set the endpoint. For existing users, if they want a different endpoint then they have already modified the YAML, so I don't think it affects them. For new users, we want them to modify the YAML instead of using the gate. So I don't think we should add this capability.

mx-psi commented:
Adding some people that commented on the original issue as reviewers :) PTAL!

jpkrohling commented on receiver/otlpreceiver/README.md:
```diff
@@ -30,7 +30,9 @@ The following settings are configurable:
-  described at https://github.com/grpc/grpc/blob/master/doc/naming.md.
+  described at https://github.com/grpc/grpc/blob/master/doc/naming.md. The 
+  `component.UseLocalHostAsDefaultHost` feature gate changes these to localhost:4317 and 
+  localhost:4318 respectively.
```
Should it say that this will be the default behavior in the future?

mx-psi replied:
Done!

jpkrohling commented on receiver/otlpreceiver/otlp.go:
```diff
@@ -163,6 +166,9 @@ func (r *otlpReceiver) startProtocolServers(host component.Host) error {
 // it also enables the metrics receiver too.
 func (r *otlpReceiver) Start(_ context.Context, host component.Host) error {
+	onceLogLocalHost.Do(func() {
+		component.LogAboutUseLocalHostAsDefault(r.settings.Logger)
```
No big deal, but I think I would log it once per receiver instance instead of once per receiver type. The reason is that the r.settings.Logger has the component name in it, which is specific to the receiver instance.

jpkrohling replied:
Alternative: if `component.LogAboutUseLocalHostAsDefault` is always logging a message anyway (see my other comment there), then the sync.Once should be at the feature flag level, and not per component?

jpkrohling commented on component/featuregate.go:
```diff
@@ -0,0 +1,40 @@
+// LogAboutUseLocalHostAsDefault logs about the upcoming change from 0.0.0.0 to localhost on server-like components.
+func LogAboutUseLocalHostAsDefault(logger *zap.Logger) {
+	if !UseLocalHostAsDefaultHostfeatureGate.IsEnabled() {
+		logger.Info(
```
I think this should only log something if the config value is set to empty or to 0.0.0.0 (current default).

mx-psi commented:
@jpkrohling after the last commit we now warn per component and only if the endpoint is not set. The approach is similar to what I did for the `logging` exporter: https://github.com/open-telemetry/opentelemetry-collector/blob/5c225d9c11ad55680f93267a169c835969b9e61f/exporter/loggingexporter/config.go#L41-L42

bogdandrutu commented on component/featuregate.go:
I don't think we intend to have "feature-gate" shared code-wise that much. Definitely a 1.0 package is not the right place, but I don't have a better idea. Can we have 2 different gates for core vs contrib so that will be an internal package?

mx-psi replied:
I would like to avoid two feature gates if possible, I think it could be confusing. One alternative to move to internal is to try to register on both core and contrib and ignore errors if related to already being registered:

```go
var UseLocalHostAsDefaultHostfeatureGate featuregate.Gate

func init() {
   var err error
   UseLocalHostAsDefaultHostfeatureGate, err = featuregate.GlobalRegistry().Register(...)
    
   if !errors.Is(err, featuregate.ErrAlreadyRegistered) { // new error type, could also just check error message
       panic(err)
   }
}
```

Does this sound okay?

bogdandrutu replied:
I like this, the setting the address code will be duplicated but I think that is fine.

bogdandrutu reviewed (approved):
I just have concern with sharing the gate (code) and then removing the code.

mx-psi commented:
@bogdandrutu see https://github.com/open-telemetry/opentelemetry-collector/pull/8622#discussion_r1374312483 for an alternative way of doing this

bogdandrutu commented on receiver/otlpreceiver/otlp.go:
```diff
@@ -44,6 +44,8 @@ type otlpReceiver struct {
 
 	settings *receiver.CreateSettings
+
+  onceLogLocalHost sync.Once
```
strange indentation?

mx-psi replied:
Fixed

bogdandrutu commented on component/featuregate.go:
```diff
@@ -0,0 +1,40 @@
+
+// EndpointForPort gets the endpoint for a given port using localhost or 0.0.0.0 depending on the feature gate.
+func EndpointForPort(port int) string {
+	host := "0.0.0.0"
```
Just to make sure we "encourage" the right before, what about you start with "localhost" and change if the gate is disabled.

mx-psi commented:
## Summary of major changes to this PR 

1. Added `featuregate.ErrIsAlreadyRegistered` to the `featuregate` module to be able to check if a gate is already registered.
2. Moved the gate definition from `component` to a new `internal/localhostgate` package. This package will be duplicated in  contrib. This avoids having this as part of the public API.
3. **Kept** the feature gate in the `component.` namespace. I can't think of a better name, and this one seems good enough.

mx-psi commented:
@bogdandrutu final look?

bogdandrutu commented on internal/localhostgate/featuregate.go:
```diff
@@ -0,0 +1,80 @@
+	gate, err := reg.Register(id, stage, opts...)
+
+	if err != nil {
+		switch {
```
nit: personally I don't like switch/one cond/default instead of `if`.

bogdandrutu commented on receiver/otlpreceiver/otlp.go:
```diff
@@ -163,6 +166,11 @@ func (r *otlpReceiver) startProtocolServers(host component.Host) error {
 // Start runs the trace receiver on the gRPC server. Currently
 // it also enables the metrics receiver too.
 func (r *otlpReceiver) Start(_ context.Context, host component.Host) error {
+	r.onceLogLocalHost.Do(func() {
```
What is the purpose of this log? Why not log once per process if the gate is disabled?

mx-psi replied:
Now logging once when the Collector starts

mx-psi commented:
@bogdandrutu (and others) will merge this by Wednesday EU morning unless somebody objects before then :)
